File: patterns/02-reliability-patterns/02-retry/python/retry.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
class FakeScheduler:
    def __init__(self):self.delays=[]
    def wait(self,value):self.delays.append(value)
def evaluate(input_data):
    scheduler=FakeScheduler();attempts=0;outcomes=input_data.get("outcomes",[])
    while attempts<input_data["max_attempts"]:
        outcome=outcomes[attempts] if attempts<len(outcomes) else "transient";attempts+=1
        if outcome=="success":return{"status":"success","attempts":attempts,"delays_ms":scheduler.delays}
        if outcome=="permanent":raise PatternError("non_retryable")
        if outcome!="transient":raise PatternError("unknown_outcome")
        if attempts<input_data["max_attempts"]:scheduler.wait(input_data["base_delay_ms"]*(2**(attempts-1)))
    return{"status":"exhausted","attempts":attempts,"delays_ms":scheduler.delays}

File: patterns/02-reliability-patterns/02-retry/go/retry.go
package retry

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type Scheduler struct{ delays []int }

func (s *Scheduler) Wait(value int) { s.delays = append(s.delays, value) }
func Evaluate(input map[string]any) (any, error) {
	scheduler := &Scheduler{delays: []int{}}
	raw, _ := input["outcomes"].([]any)
	maxAttempts := int(input["max_attempts"].(float64))
	base := int(input["base_delay_ms"].(float64))
	attempts := 0
	for attempts < maxAttempts {
		outcome := "transient"
		if attempts < len(raw) {
			outcome = raw[attempts].(string)
		}
		attempts++
		switch outcome {
		case "success":
			return map[string]any{"status": "success", "attempts": attempts, "delays_ms": scheduler.delays}, nil
		case "permanent":
			return nil, &PatternError{code: "non_retryable"}
		case "transient":
		default:
			return nil, &PatternError{code: "unknown_outcome"}
		}
		if attempts < maxAttempts {
			scheduler.Wait(base << (attempts - 1))
		}
	}
	return map[string]any{"status": "exhausted", "attempts": attempts, "delays_ms": scheduler.delays}, nil
}
